File: chess_cpp/statecheckers/BaseStateChangeChecker.cpp
//
// Created by Hrach Khachatryan on 12/16/19.
//

#include "BaseStateChangeChecker.h"

void BaseStateChangeChecker::addChecker(BaseStateChangeChecker *checker) {
    if (nextChecker == nullptr) {
        nextChecker = checker;
    } else {
        nextChecker->addChecker(checker);
    }
}

void BaseStateChangeChecker::execute(Board *board, Spot *start, Spot *end, std::vector<std::string> messages) {
    check(board, start, end, messages);
    if (nextChecker != nullptr)
        nextChecker->execute(board, start, end, messages);
}

File: chess_java/src/chess/statecheckers/BaseStateChangeChecker.java
package chess.statecheckers;

import chess.Board;
import chess.Spot;

import java.util.List;

public abstract class BaseStateChangeChecker {

    private BaseStateChangeChecker nextChecker;

    public void addChecker(BaseStateChangeChecker checker) {
        if (nextChecker == null) {
            nextChecker = checker;
        } else {
            nextChecker.addChecker(checker);
        }
    }

    public void execute(Board board, Spot start, Spot end, List<String> messages){
        check(board, start, end, messages);
        if(nextChecker != null)
            nextChecker.execute(board, start, end, messages);
    }

    public abstract void check(Board board, Spot start, Spot end, List<String> messages) ;

}
